gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java

```diff
@@ -19,126 +19,81 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 
+import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableMap;
 import java.util.Map;
+import javax.annotation.Nullable;
 
-/**
- * Options for creating objects in GCS.
- */
-public class CreateObjectOptions {
-  public static final ImmutableMap<String, byte[]> EMPTY_METADATA = ImmutableMap.of();
-  public static final String DEFAULT_CONTENT_TYPE = "application/octet-stream";
-  public static final String DEFAULT_CONTENT_ENCODING = null;
-  public static final CreateObjectOptions DEFAULT =
-      new CreateObjectOptions(/* overwriteExisting= */ true);
-
-  private final boolean overwriteExisting;
-  private final String contentType;
-  private final String contentEncoding;
-  private final Map<String, byte[]> metadata;
-  private final boolean requireMetadataMatchForEmptyObjects;
+/** Options that can be specified when creating a file in the {@link GoogleCloudStorage}. */
+@AutoValue
+public abstract class CreateObjectOptions {
 
-  /**
-   * Construct a new CreateObjectOptions with empty metadata and the default content-type.
-   * Since no metadata was explicitly provided, we will set metadata-match for
-   * empty objects to false.
-   *
-   * @param overwriteExisting True to overwrite any existing objects with the same name.
-   */
-  public CreateObjectOptions(boolean overwriteExisting) {
-    this(overwriteExisting, DEFAULT_CONTENT_TYPE, DEFAULT_CONTENT_ENCODING, EMPTY_METADATA, false);
-  }
+  public static final String CONTENT_TYPE_DEFAULT = "application/octet-stream";
 
-  /**
-   * Construct a new CreateObjectOptions with the specified metadata, and default content-type.
-   * Since metadata was explicitly provided, even if empty, we will set metadata-match for
-   * empty objects to true.
-   * 
-   *
-   * @param overwriteExisting True to overwrite any existing objects with the same name.
-   * @param metadata A dictionary of metadata to apply to created objects.
-   */
-  public CreateObjectOptions(boolean overwriteExisting, Map<String, byte[]> metadata) {
-    this(overwriteExisting, DEFAULT_CONTENT_TYPE, metadata);
-  }
+  public static final CreateObjectOptions DEFAULT_NO_OVERWRITE = builder().build();
```

Comment:
Should we still keep a general DEFAULT option here? Am thinking not all methods require the overwrite semantic. Or you think it's better to explicitly indicate overwrite or non-overwrite in default? It's just feel weird that in context that a real default option is required and DEFAULT_OVERWRITE is used. One option I can think of is having a DEFAULT (with overwirte set to true) and a OPTION_NO_OVERWRITE(with no default semantic in it)

Reply:
Yes, I want to be explicit because it's quite important configuration that can cause data loss.